File: api/src/main/java/net/blueshell/api/model/Committee.java
package net.blueshell.api.model;

import jakarta.persistence.*;
import lombok.Data;
import net.blueshell.api.base.BaseModel;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;

import java.sql.Timestamp;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

@Entity
@Table(name = "committees")
@Data
@SQLDelete(sql = "UPDATE committees SET deleted_at = CURRENT_TIMESTAMP WHERE id = ?")
@SQLRestriction("deleted_at >= NOW()")
public class Committee implements BaseModel {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "name")
    private String name;

    @Column(name = "description")
    private String description;

    @OneToMany(
            mappedBy = "committee",
            cascade = CascadeType.ALL,
            orphanRemoval = true,
            fetch = FetchType.EAGER)
    private Set<CommitteeMember> members = new HashSet<>();

    @Column(name = "deleted_at")
    private Timestamp deletedAt;

    public Committee() {
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Committee committee = (Committee) o;
        return id == committee.id;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    public boolean hasMember(User user) {
        return getMembers().stream().anyMatch(cm -> cm.getUser().getId().equals(user.getId()));
    }

     public Set<User> getUsers() {
         return members == null ? Set.of() :
             members.stream().map(CommitteeMember::getUser).filter(Objects::nonNull).collect(Collectors.toSet());
     }
}

File: api/src/main/kotlin/net/blueshell/api/committee/persistence/Committee.kt
package net.blueshell.api.committee.persistence

import jakarta.persistence.*
import net.blueshell.api.shared.model.AuditedAutoIdEntity
import net.blueshell.api.user.persistence.User
import org.hibernate.annotations.SQLDelete
import org.hibernate.annotations.SQLRestriction

@Entity
@Table(
    name = "committees",
    uniqueConstraints = [
        UniqueConstraint(name = "uk_committees_name_deleted_at", columnNames = ["name", "deleted_at"]),
    ],
    indexes = [
        Index(name = "idx_committees_deleted_at", columnList = "deleted_at"),
        Index(name = "idx_committees_name", columnList = "name"),
    ]
)
@SQLDelete(sql = "UPDATE committees SET deleted_at = NOW(), version = version + 1 WHERE id = ? AND version = ?")
@SQLRestriction("deleted_at = '9999-12-31 23:59:59'")
class Committee : AuditedAutoIdEntity() {
    @Column(name = "name", nullable = false)
    lateinit var name: String

    @Column(name = "description", nullable = false, length = 4095)
    lateinit var description: String

    @OneToMany(mappedBy = "committee", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.LAZY)
    private val _members: MutableList<CommitteeMember> = mutableListOf()
    val members: List<CommitteeMember>
        get() = _members

    fun hasMember(user: User?): Boolean {
        return user != null && _members.any { cm -> cm.user.id == user.id }
    }

    fun replaceMembers(members: List<CommitteeMember>) {
        _members.clear()
        _members.addAll(members)
        _members.forEach { it.committee = this }
    }
}
